File: examples/src/main/java/com/teamdev/jxbrowser/examples/swing/CustomContextMenu.java
/*
 *  Copyright 2025, TeamDev. All rights reserved.
 *
 *  Redistribution and use in source and/or binary forms, with or without
 *  modification, must retain the above copyright notice and the following
 *  disclaimer.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.teamdev.jxbrowser.examples.swing;

import static com.teamdev.jxbrowser.engine.RenderingMode.HARDWARE_ACCELERATED;
import static javax.swing.SwingUtilities.invokeLater;

import com.teamdev.jxbrowser.browser.callback.ShowContextMenuCallback;
import com.teamdev.jxbrowser.browser.internal.BrowserImpl;
import com.teamdev.jxbrowser.engine.Engine;
import com.teamdev.jxbrowser.engine.RenderingMode;
import com.teamdev.jxbrowser.frame.Frame;
import com.teamdev.jxbrowser.internal.rpc.stream.Interceptor.Action;
import com.teamdev.jxbrowser.ui.event.internal.rpc.MoveMouseWheel;
import com.teamdev.jxbrowser.view.swing.BrowserView;
import java.awt.BorderLayout;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import javax.swing.JFrame;
import javax.swing.JMenuItem;
import javax.swing.JPopupMenu;
import javax.swing.WindowConstants;
import javax.swing.event.PopupMenuEvent;
import javax.swing.event.PopupMenuListener;

/**
 * This example demonstrates how to build a custom context menu using Swing API and show it when
 * user right-click on the web page.
 */
public final class CustomContextMenu {

    public static void main(String[] args) {
        var engine = Engine.newInstance(HARDWARE_ACCELERATED);
        var browser = engine.newBrowser();

        invokeLater(() -> {
            var view = BrowserView.newInstance(browser);
            browser.set(ShowContextMenuCallback.class, new ShowCustomContextMenu(view));

            var frame = new JFrame("JxBrowser: Swing Context Menu");
            frame.addWindowListener(new WindowAdapter() {
                @Override
                public void windowClosing(WindowEvent e) {
                    engine.close();
                }
            });
            frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
            frame.add(view, BorderLayout.CENTER);
            frame.setSize(900, 700);
            frame.setLocationRelativeTo(null);
            frame.setVisible(true);
        });

        browser.navigation().loadUrl("https://html5test.teamdev.com/");
    }

    private static class ShowCustomContextMenu implements ShowContextMenuCallback {

        private final BrowserView parent;

        ShowCustomContextMenu(BrowserView parent) {
            this.parent = parent;
        }

        @Override
        public void on(Params params, Action action) {
            new SwingContextMenu(parent, params, action) {
                @Override
                protected void initialize(JPopupMenu contextMenu) {
                    var browser = parent.getBrowser();

                    var goBackMenuItem = new JMenuItem("Back");
                    goBackMenuItem.addActionListener(
                            e -> browser.navigation().goBack());

                    var goForwardMenuItem = new JMenuItem("Forward");
                    goForwardMenuItem.addActionListener(
                            e -> browser.navigation().goForward());

                    var reloadMenuItem = new JMenuItem("Reload");
                    reloadMenuItem.addActionListener(
                            e -> browser.navigation().reload());

                    var printMenuItem = new JMenuItem("Print...");
                    printMenuItem.addActionListener(
                            e -> browser.mainFrame().ifPresent(Frame::print));

                    contextMenu.add(goBackMenuItem);
                    contextMenu.add(goForwardMenuItem);
                    contextMenu.add(reloadMenuItem);
                    contextMenu.addSeparator();
                    contextMenu.add(printMenuItem);
                }
            }.show();
        }
    }

    private static abstract class SwingContextMenu {

        private final BrowserView parent;
        private final BrowserImpl browser;
        protected final ShowContextMenuCallback.Params params;
        protected final ShowContextMenuCallback.Action callback;

        SwingContextMenu(BrowserView parent, ShowContextMenuCallback.Params params,
                ShowContextMenuCallback.Action callback) {
            this.parent = parent;
            this.browser = (BrowserImpl) parent.getBrowser();
            this.params = params;
            this.callback = callback;
        }

        protected abstract void initialize(JPopupMenu contextMenu);

        public final void show() {
            var popupMenu = new JPopupMenu();
            popupMenu.setLightWeightPopupEnabled(false);
            popupMenu.addPopupMenuListener(
                    new SwingContextMenu.PopupMenuListenerImpl());
            initialize(popupMenu);
            registerMouseCallbackInterceptors(popupMenu);
            var location = params.location();
            int x = location.x();
            int y = location.y();
            invokeLater(() -> popupMenu.show(parent, x, y));
        }

        private void notifyContextMenuClosed() {
            unregisterMouseCallbackInterceptors();
            if (!callback.isClosed()) {
                callback.close();
            }

        }

        private void registerMouseCallbackInterceptors(JPopupMenu popupMenu) {
            if (isHardwareAccelerated()) {
                browser.setCallbackInterceptor(MoveMouseWheel.Request.class, (params) -> {
                    hide(popupMenu);
                    return Action.PROCEED;
                });
            }

        }

        private void unregisterMouseCallbackInterceptors() {
            if (isHardwareAccelerated()) {
                browser.setCallbackInterceptor(MoveMouseWheel.Request.class,
                        (params) -> Action.PROCEED);
            }

        }

        private void hide(JPopupMenu popupMenu) {
            invokeLater(() -> {
                if (popupMenu.isShowing()) {
                    popupMenu.setVisible(false);
                }
            });
        }

        private boolean isHardwareAccelerated() {
            return browser.engine().options().renderingMode()
                    == RenderingMode.HARDWARE_ACCELERATED;
        }

        private class PopupMenuListenerImpl implements PopupMenuListener {

            private PopupMenuListenerImpl() {
            }

            public void popupMenuWillBecomeVisible(PopupMenuEvent e) {
            }

            public void popupMenuWillBecomeInvisible(PopupMenuEvent e) {
                invokeLater(SwingContextMenu.this::notifyContextMenuClosed);
            }

            public void popupMenuCanceled(PopupMenuEvent e) {
            }
        }
    }
}

File: examples/src/main/kotlin/com/teamdev/jxbrowser/examples/compose/CustomContextMenu.kt
/*
 *  Copyright 2026, TeamDev. All rights reserved.
 *
 *  Redistribution and use in source and/or binary forms, with or without
 *  modification, must retain the above copyright notice and the following
 *  disclaimer.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.teamdev.jxbrowser.examples.compose

import androidx.compose.material.CursorDropdownMenu
import androidx.compose.material.DropdownMenuItem
import androidx.compose.material.Text
import androidx.compose.runtime.*
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.window.singleWindowApplication
import com.teamdev.jxbrowser.browser.Browser
import com.teamdev.jxbrowser.browser.callback.ShowContextMenuCallback
import com.teamdev.jxbrowser.dsl.Engine
import com.teamdev.jxbrowser.dsl.browser.mainFrame
import com.teamdev.jxbrowser.dsl.browser.navigation
import com.teamdev.jxbrowser.dsl.register
import com.teamdev.jxbrowser.dsl.removeCallback
import com.teamdev.jxbrowser.engine.RenderingMode
import com.teamdev.jxbrowser.view.compose.BrowserView

/**
 * This example demonstrates how to create a custom context menu using
 * [ShowContextMenuCallback] and Compose API.
 *
 * [Browser] invokes this callback when a user right-clicks on a web page.
 * We are going to create our own [CursorDropdownMenu] and bind it to
 * the browser with the callback.
 */
fun main() = singleWindowApplication(title = "Custom context menu") {
    val engine = remember { Engine(RenderingMode.OFF_SCREEN) }
    val browser = remember { engine.newBrowser() }

    // Store dropdown menu state.
    var menuShowed by remember { mutableStateOf(false) }
    val clipboard = LocalClipboardManager.current

    // Add browser view.
    BrowserView(browser)

    // Add custom context menu.
    CursorDropdownMenu(
        expanded = menuShowed, // Controls menu's visibility.
        onDismissRequest = { menuShowed = false },
    ) {
        DropdownMenuItem(onClick = {
            val selectedText = browser.mainFrame?.selectionAsText()!!
            clipboard.setText(AnnotatedString(selectedText))
            menuShowed = false
        }
        ) {
            Text("Copy")
        }
        DropdownMenuItem(onClick = {
            browser.navigation.goBack()
            menuShowed = false
        }) {
            Text("Back")
        }
        DropdownMenuItem(onClick = {
            browser.navigation.goForward()
            menuShowed = false
        }) {
            Text("Forward")
        }
        DropdownMenuItem(onClick = {
            browser.navigation.reload()
            menuShowed = false
        }) {
            Text("Reload")
        }
    }

    DisposableEffect(Unit) {

        // Register a callback to display the menu
        // upon right-clicking on a web page.
        browser.register(ShowContextMenuCallback { _, tell ->
            menuShowed = true
            tell.close() // We don't need its actions.
        })

        browser.navigation().loadUrl("https://html5test.teamdev.com/")

        onDispose {

            // Remove the callback when it is no longer needed.
            browser.removeCallback<ShowContextMenuCallback>()

            engine.close()
        }
    }
}
